{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Variant Calling\n",
    "\n",
    "Now that we have the aligned sequences to the reference, we can look for difference between the aligned sequences and the reference. This process is also known as variant calling.\n",
    "\n",
    "In order to perform variant calling, we need to preprocess the SAM file as well as the reference chromosome 5 fasta file. The following steps will be performed :\n",
    "\n",
    "* Converting SAM to BAM format (this is the binary compressed version)\n",
    "* Sorting the BAM file\n",
    "* Indexing the BAM file\n",
    "* Indexing the reference fasta file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Processing the SAM file generated by Bowtie2\n",
    "\n",
    "We will process the SAM file using samtools, which can be accessed after loading the module."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will begin by converting the SAM file to the compressed BAM format. To do this, we will need to refer to the reference fasta file and the SAM file. The output will be redirected to a new BAM file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samtools view -Sb SRR12165154.sam -o SRR12165154.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice that the BAM files is much smaller than the SAM file\n",
    "\n",
    "Next, we will sort and index the BAM file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samtools sort SRR12165154.bam -o SRR12165154_sorted.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samtools index SRR12165154_sorted.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Indexing the reference fasta file\n",
    "\n",
    "To prepare the reference fasta file for variant calling, we need to index the file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samtools faidx Homo_sapiens.GRCh38.dna.primary_assembly.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Variant calling using Bcftools\n",
    "\n",
    "Bcftools is a utilities for variant calling and manipulating VCFs and BCFs.\n",
    "http://samtools.github.io/bcftools/bcftools.html\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bcftools mpileup -ou -f Homo_sapiens.GRCh38.dna.primary_assembly.fa SRR12165154_sorted.bam -d 80 -o SRR12165154_output.pileup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bcftools call -mv -Ov SRR12165154_output.pileup --ploidy 1 -o SRR12165154_snp.vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### VCF Format\n",
    "\n",
    "The Variant Call Format (VCF) is a text-based format for specifying variants. An example is shown below:\n",
    "\n",
    "![image](https://raw.githubusercontent.com/fikaparamita/var-calling/main/images/vcfexample.png?token=ARW6RR7SE67JOD742UJ6CZS7VSB5I)\n",
    "\n",
    "\n",
    "The basic fields are as follows:\n",
    "\n",
    "![image](https://raw.githubusercontent.com/fikaparamita/var-calling/main/images/vcf.png?token=ARW6RR3E5QW7527EJ26H2CC7VSB3K)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": ""
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
